src/com/reco1l/osu/hud/HUDElement.kt

```diff
@@ -0,0 +1,303 @@
+            setPosition(data.position.x, data.position.y)
+        }
+    }
+
+    open fun getSkinData() = HUDElementSkinData(
+        type = this::class,
+        anchor = anchor,
+        origin = origin,
+        scale = scaleX, // Scale is uniform currently.
+        position = Vec2(x, y),
+        rotation = rotation
+    )
+
+    //endregion
+
+    //region Element events
+
+    open fun setEditMode(value: Boolean) {
+        isInEditMode = value
+
+        if (value) {
+            background = HUDElementBackground()
+            overlay = HUDElementOverlay(this)
+
+            parent!!.attachChild(overlay!!)
+        } else {
+            connectionLine?.detachSelf()
+            overlay?.detachSelf()
+
+            connectionLine = null
+            background = null
+            overlay = null
+        }
+    }
+
+    open fun onSelectionStateChange(isSelected: Boolean) {
+
+        (background as? HUDElementBackground)?.isSelected = isSelected
+        overlay?.isVisible = isSelected
+
+        if (isSelected) {
+            updateConnectionLine()
+            connectionLine?.isVisible = true
+        } else {
+            connectionLine?.isVisible = false
+        }
+    }
+
+    //endregion
+
+    //region Entity events
+
+    private var initialX = 0f
+    private var initialY = 0f
+
+    override fun onAreaTouched(event: TouchEvent, localX: Float, localY: Float): Boolean {
+
+        if (isInEditMode) {
+            val parentLocalX = drawX + localX
+            val parentLocalY = drawY + localY
+
+            if (event.isActionDown) {
+                (parent as? GameplayHUD)?.selected = this
+
+                initialX = parentLocalX
+                initialY = parentLocalY
+                return true
+            }
+
+            if (event.isActionMove) {
+                val deltaX = parentLocalX - initialX
+                val deltaY = parentLocalY - initialY
+
+                // Preventing from moving the element if it's not selected.
+                if ((parent as? GameplayHUD)?.selected == this) {
+                    applyClosestAnchorOrigin()
+                    x += deltaX
+                    y += deltaY
+                    updateConnectionLine()
+
+                    initialX = parentLocalX
+                    initialY = parentLocalY
+                    return true
+                }
+            }
+
+            if (event.isActionUp) {
+                return (parent as? GameplayHUD)?.selected == this
+            }
+        }
+
+        return false
+    }
+
+    override fun invalidateTransformations() {
+        super.invalidateTransformations()
+        overlay?.invalidateTransformations()
+    }
+
+
+    private fun applyClosestAnchorOrigin() {
+
+        val drawSize = drawSize
+        val drawPosition = drawPosition
+        val parentDrawSize = (parent as ExtendedEntity).drawSize
+
+        val relativeTopLeftDrawPosition = (drawPosition + drawSize / 2f) / parentDrawSize
```

Comment:
Shouldn't the relative top-left position be just `drawPosition / parentDrawSize`? With the addition of `drawSize / 2`, this looks like the relative *center* position.